### ISSUE TITLE
git_status_list is slower than `git status`

### ISSUE THREAD

**Issue by powerman** [libgit2/libgit2#4230]
### Reproduction steps

Build provided `libgit2status` tool (I'll post it source in the next comment to keep initial report easier to read) and run it in the root of large enough repo. Compare it speed with `git status`.

```
% go version
go version go1.8.1 linux/amd64
% git --version
git version 2.10.2
% git clone https://github.com/gentoo/gentoo.git
% cd gentoo
% find -type f | wc -l
99540
% find -type d | wc -l
27306
% git config core.untrackedCache false
% time git status --short

0.257 real  0.128 user  0.206 sys  32MB RAM
% git config core.untrackedCache true
% time git status --short

0.097 real  0.067 user  0.110 sys  38MB RAM
% time libgit2status
{AddedFiles:0 ModifiedFiles:0 DeletedFiles:0 RenamedFiles:0 UnmergedFiles:0 HasUntrackedFiles:false}

0.626 real  0.340 user  0.282 sys  154MB RAM
% touch profiles/base/testfile.added
% touch profiles/base/testfile.new
% git add profiles/base/testfile.added
% time git status --short
A  profiles/base/testfile.added
?? profiles/base/testfile.new

0.115 real  0.082 user  0.108 sys  38MB RAM
% git config core.untrackedCache false
% time git status --short
A  profiles/base/testfile.added
?? profiles/base/testfile.new

0.255 real  0.134 user  0.193 sys  32MB RAM
% time libgit2status
{AddedFiles:1 ModifiedFiles:0 DeletedFiles:0 RenamedFiles:0 UnmergedFiles:0 HasUntrackedFiles:true}

0.611 real  0.358 user  0.250 sys  154MB RAM
```

### Expected behavior

Well… in my dream git2go should works at least as fast as `git status`… actually faster if we just needs to know is workdir is dirty or is it has any untracked files - without needs to know full list of all added/modified/untracked/etc. files output by `git status`.

### Actual behavior

It's nearly 2.5 times slower than `git status` on my system (HDD, i7-2600K overcloked @ 4.5GHz) without untrackedCache and 5-6 times slower with untrackedCache.

While strace-ing `git status` I've noticed it calls `lstat` from 2 or 3 threads - can this be the reason?

For all `time`d runs above I take average result from several runs and my system has enough RAM to keep dircache for the full repo, so there was no (noticeable, at least) disk accesses, and thus HDD/SSD shouldn't make any difference.

### Version of libgit2 (release number or SHA1)

df4dfaa

This version is vendored by git2go. I've tried to update to current master but get this error:
```
./submodule.go:339: ptr.clone_checkout_strategy undefined (type *C.struct_git_submodule_update_options has no field or method clone_checkout_strategy)
```
If something was just renamed then it probably can be easily fixed to allow testing with latest master.

### Operating system(s) tested

Hardened Gentoo Linux, amd64.

**Comment by powerman** [comment]
Build it with `go install -tags static` or `go install -tags static -buildmode pie` (latest is needed on systems with PaX-patched kernel).
```go
package main

import (
	"fmt"
	"log"

	git2go "github.com/libgit2/git2go"
)

type RepoStatus struct {
	AddedFiles        int
	ModifiedFiles     int
	DeletedFiles      int
	RenamedFiles      int
	UnmergedFiles     int
	HasUntrackedFiles bool
}

func main() {
	fmt.Printf("%+v\n", gitStatus())
}

func gitStatus() (status RepoStatus) {
	// XXX Do not call Free() on any object - this should be faster
	// and safe for short-lived command. DO NOT COPY&PASTE THIS AS IS!

	repo, err := git2go.OpenRepository(".")
	if err != nil {
		return // not a (valid) repo
	}

	statuses, err := repo.StatusList(&git2go.StatusOptions{
		Show: git2go.StatusShowIndexAndWorkdir,
		Flags: git2go.StatusOptUpdateIndex |
			git2go.StatusOptIncludeUntracked |
			git2go.StatusOptRenamesHeadToIndex |
			git2go.StatusOptExcludeSubmodules,
		Pathspec: nil,
	})
	if err != nil {
		log.Println("repo.StatusList:", err)
		return
	}
	n, err := statuses.EntryCount()
	if err != nil {
		log.Println("statuses.EntryCount:", err)
		return
	}

	for i := 0; i < n; i++ {
		entry, err := statuses.ByIndex(i)
		if err != nil {
			log.Println("statuses.ByIndex:", err)
			return
		}
		if entry.Status == git2go.StatusCurrent { // == 0, so must check it first
			// never here: StatusOptIncludeUnmodified wasn't enabled
			log.Println("entry.Status=unmodified")
			return
		}
		if entry.Status&git2go.StatusIndexNew > 0 {
			entry.Status &^= git2go.StatusIndexNew
			status.AddedFiles++
		}
		if entry.Status&(git2go.StatusIndexModified|git2go.StatusWtModified) > 0 {
			entry.Status &^= git2go.StatusIndexModified | git2go.StatusWtModified
			status.ModifiedFiles++
		}
		if entry.Status&(git2go.StatusIndexDeleted|git2go.StatusWtDeleted) > 0 {
			entry.Status &^= git2go.StatusIndexDeleted | git2go.StatusWtDeleted
			status.DeletedFiles++
		}
		if entry.Status&(git2go.StatusIndexRenamed|git2go.StatusWtRenamed) > 0 {
			entry.Status &^= git2go.StatusIndexRenamed | git2go.StatusWtRenamed
			status.RenamedFiles++
		}
		if entry.Status&(git2go.StatusIndexTypeChange|git2go.StatusWtTypeChange) > 0 {
			entry.Status &^= git2go.StatusIndexTypeChange | git2go.StatusWtTypeChange
			status.ModifiedFiles++
		}
		switch entry.Status {
		case 0: // was bitmask flag(s)
		case git2go.StatusWtNew:
			status.HasUntrackedFiles = true
		case git2go.StatusIgnored:
			continue
		case git2go.StatusConflicted:
			status.UnmergedFiles++
		default:
			// May be "unreadable" - this const wasn't
			// imported from C, not sure why.
			log.Println("entry.Status unknown:", entry.Status, entry)
			continue
		}
	}

	return
}
```

**Comment by powerman** [comment]
I was looking at git2go/libgit2 to implement git status at shell $PS1 prompt, and thus delay more than about 0.1sec is really noticeable and critical (for this task). So for now only real option I've is to exec `git status` with untrackedCache enabled. But I'll be happy to try libgit2 again when it'll became as fast as `git status` (or if this issue somehow is in my code and not in git2go/libgit2).

**Comment by mathomp4** [comment]
As Github has kindly pointed out, I might have encountered this as well. On a repo of mine `git status` takes ~4 seconds which `git-status-prompt` which is mainly calls to libgit2, takes 46 seconds.

**Comment by baldurk** [comment]
I ran into this after investigating slowness in gitkraken which calls `git_status_list_new` often (and synchronously on the UI - boo).

I found that a lot of time is spent in `GetFileAttributesExW` which is used for the implementation of `lstat_w`. From an uneducated standpoint, it seems like possibly caching the results of that more aggressively might help (it's called more than once for each file).

I also noticed that for every file there's a search for `.gitattributes` in each directory up to the repo root, and I think similarly for `.gitignore`. I guess this is intended behaviour to allow overrides, but I'd wager that it's quite rare to find those files anywhere other than the repo root, so it might be nice to have a way to toggle this off with a gitconfig setting.

Even if not, it doesn't seem to cache the results of searching for `.gitattributes` everywhere which means that deep directory structures with lots of files can hurt quite badly re-checking for non-existent files over and over.

For anyone else interested, I was able to improve matters a little bit by judicious use of `git update-index --assume-unchanged` which brought the perf hit down enough that I can ignore this problem for now.

**Comment by ethomson** [comment]
I'm surprised by this.  I'm surprised that we're calling `GetFileAttributesEx` at all - we should be using the file attributes from `FindFirstFile`.  We should also be caching the results of attributes lookups.

I'm not sure what's amiss here.  I'm not particularly excited about a configuration setting that changes the way we look for `.gitattributes` or `.gitignore` - at least not as a solution for this problem, as that seems like a weak bandaid.

**Comment by baldurk** [comment]
That's totally fair, I am not familiar with libgit2's internals so from the outside it just seemed like a low-value high-cost operation that might be useful to disable on a toggle.

For what it's worth from my debugging here are some of the stacks where it seems to hit `GetFileAttributesEx` on a build of e212011b9872c52f6205d3a30b10f753c3108918:

<details><summary>Callstack 1 (`path = "path/to/some/file"`)</summary>
<p>

```
>	git2.dll!lstat_w(wchar_t * path, p_stat * buf, unsigned char posix_enotdir) Line 303	C
 	git2.dll!p_stat(const char * path, p_stat * buf) Line 691	C
 	git2.dll!git_path_isdir(const char * path) Line 553	C
 	git2.dll!git_attr_path__init(git_attr_path * info, const char * path, const char * base, git_dir_flag dir_flag) Line 512	C
 	git2.dll!git_attr_get_many_with_session(const char * * values, git_repository * repo, git_attr_session * attr_session, unsigned int flags, const char * pathname, unsigned __int64 num_attr, const char * * names) Line 131	C
 	git2.dll!filter_list_check_attributes(const char * * * out, git_repository * repo, git_attr_session * attr_session, git_filter_def * fdef, const git_filter_source * src) Line 436	C
 	git2.dll!git_filter_list__load_ext(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, git_filter_options * filter_opts) Line 522	C
 	git2.dll!git_filter_list_load(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, unsigned int flags) Line 588	C
 	git2.dll!git_diff__oid_for_entry(git_oid * out, git_diff * d, const git_index_entry * src, unsigned short mode, const git_oid * update_match) Line 634	C
 	git2.dll!maybe_modified(git_diff_generated * diff, diff_in_progress * info) Line 847	C
 	git2.dll!handle_matched_item(git_diff_generated * diff, diff_in_progress * info) Line 1172	C
 	git2.dll!git_diff__from_iterators(git_diff * * out, git_repository * repo, git_iterator * old_iter, git_iterator * new_iter, const git_diff_options * opts) Line 1245	C
 	git2.dll!git_diff_index_to_workdir(git_diff * * out, git_repository * repo, git_index * index, const git_diff_options * opts) Line 1381	C
 	git2.dll!git_status_list_new(git_status_list * * out, git_repository * repo, const git_status_options * opts) Line 350	C
 	status.exe!main(int argc, char * * argv) Line 113	C
```

</p>
</details>

<details><summary>Callstack 2 (`path = ".git/info/attributes"`)</summary>
<p>

```
>	git2.dll!lstat_w(wchar_t * path, p_stat * buf, unsigned char posix_enotdir) Line 300	C
 	git2.dll!p_stat(const char * path, p_stat * buf) Line 691	C
 	git2.dll!git_attr_file__load(git_attr_file * * out, git_repository * repo, git_attr_session * attr_session, git_attr_file_entry * entry, git_attr_file_source source, int(*)(git_repository *, git_attr_file *, const char *) parser) Line 140	C
 	git2.dll!git_attr_cache__get(git_attr_file * * out, git_repository * repo, git_attr_session * attr_session, git_attr_file_source source, const char * base, const char * filename, int(*)(git_repository *, git_attr_file *, const char *) parser) Line 231	C
 	git2.dll!preload_attr_file(git_repository * repo, git_attr_session * attr_session, git_attr_file_source source, const char * base, const char * file) Line 263	C
 	git2.dll!attr_setup(git_repository * repo, git_attr_session * attr_session) Line 345	C
 	git2.dll!collect_attr_files(git_repository * repo, git_attr_session * attr_session, unsigned int flags, const char * path, git_vector * files) Line 502	C
 	git2.dll!git_attr_get_many_with_session(const char * * values, git_repository * repo, git_attr_session * attr_session, unsigned int flags, const char * pathname, unsigned __int64 num_attr, const char * * names) Line 134	C
 	git2.dll!filter_list_check_attributes(const char * * * out, git_repository * repo, git_attr_session * attr_session, git_filter_def * fdef, const git_filter_source * src) Line 436	C
 	git2.dll!git_filter_list__load_ext(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, git_filter_options * filter_opts) Line 522	C
 	git2.dll!git_filter_list_load(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, unsigned int flags) Line 588	C
 	git2.dll!git_diff__oid_for_entry(git_oid * out, git_diff * d, const git_index_entry * src, unsigned short mode, const git_oid * update_match) Line 634	C
 	git2.dll!maybe_modified(git_diff_generated * diff, diff_in_progress * info) Line 847	C
 	git2.dll!handle_matched_item(git_diff_generated * diff, diff_in_progress * info) Line 1172	C
 	git2.dll!git_diff__from_iterators(git_diff * * out, git_repository * repo, git_iterator * old_iter, git_iterator * new_iter, const git_diff_options * opts) Line 1245	C
 	git2.dll!git_diff_index_to_workdir(git_diff * * out, git_repository * repo, git_index * index, const git_diff_options * opts) Line 1381	C
 	git2.dll!git_status_list_new(git_status_list * * out, git_repository * repo, const git_status_options * opts) Line 350	C
 	status.exe!main(int argc, char * * argv) Line 113	C
```

</p>
</details>

<details><summary>Callstack 3 (`path = ".gitattributes"`)</summary>
<p>

```>	git2.dll!lstat_w(wchar_t * path, p_stat * buf, unsigned char posix_enotdir) Line 303	C
 	git2.dll!p_stat(const char * path, p_stat * buf) Line 686	C
 	git2.dll!git_futils_filestamp_check(git_futils_filestamp * stamp, const char * path) Line 1099	C
 	git2.dll!git_attr_file__out_of_date(git_repository * repo, git_attr_session * attr_session, git_attr_file * file) Line 207	C
 	git2.dll!git_attr_cache__get(git_attr_file * * out, git_repository * repo, git_attr_session * attr_session, git_attr_file_source source, const char * base, const char * filename, int(*)(git_repository *, git_attr_file *, const char *) parser) Line 230	C
 	git2.dll!preload_attr_file(git_repository * repo, git_attr_session * attr_session, git_attr_file_source source, const char * base, const char * file) Line 263	C
 	git2.dll!attr_setup(git_repository * repo, git_attr_session * attr_session) Line 348	C
 	git2.dll!collect_attr_files(git_repository * repo, git_attr_session * attr_session, unsigned int flags, const char * path, git_vector * files) Line 502	C
 	git2.dll!git_attr_get_many_with_session(const char * * values, git_repository * repo, git_attr_session * attr_session, unsigned int flags, const char * pathname, unsigned __int64 num_attr, const char * * names) Line 134	C
 	git2.dll!filter_list_check_attributes(const char * * * out, git_repository * repo, git_attr_session * attr_session, git_filter_def * fdef, const git_filter_source * src) Line 436	C
 	git2.dll!git_filter_list__load_ext(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, git_filter_options * filter_opts) Line 522	C
 	git2.dll!git_filter_list_load(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, unsigned int flags) Line 588	C
 	git2.dll!git_diff__oid_for_entry(git_oid * out, git_diff * d, const git_index_entry * src, unsigned short mode, const git_oid * update_match) Line 634	C
 	git2.dll!maybe_modified(git_diff_generated * diff, diff_in_progress * info) Line 847	C
 	git2.dll!handle_matched_item(git_diff_generated * diff, diff_in_progress * info) Line 1172	C
 	git2.dll!git_diff__from_iterators(git_diff * * out, git_repository * repo, git_iterator * old_iter, git_iterator * new_iter, const git_diff_options * opts) Line 1245	C
 	git2.dll!git_diff_index_to_workdir(git_diff * * out, git_repository * repo, git_index * index, const git_diff_options * opts) Line 1381	C
 	git2.dll!git_status_list_new(git_status_list * * out, git_repository * repo, const git_status_options * opts) Line 350	C
 	status.exe!main(int argc, char * * argv) Line 113	C
```

</p>
</details>

<details><summary>Callstack 4 (`path = "path/to/some/file"`)</summary>
<p>

```
>	git2.dll!lstat_w(wchar_t * path, p_stat * buf, unsigned char posix_enotdir) Line 303	C
 	git2.dll!p_stat(const char * path, p_stat * buf) Line 691	C
 	git2.dll!git_path_isdir(const char * path) Line 553	C
 	git2.dll!git_attr_path__init(git_attr_path * info, const char * path, const char * base, git_dir_flag dir_flag) Line 512	C
 	git2.dll!git_attr_get_many_with_session(const char * * values, git_repository * repo, git_attr_session * attr_session, unsigned int flags, const char * pathname, unsigned __int64 num_attr, const char * * names) Line 131	C
 	git2.dll!filter_list_check_attributes(const char * * * out, git_repository * repo, git_attr_session * attr_session, git_filter_def * fdef, const git_filter_source * src) Line 436	C
 	git2.dll!git_filter_list__load_ext(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, git_filter_options * filter_opts) Line 522	C
 	git2.dll!git_filter_list_load(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, unsigned int flags) Line 588	C
 	git2.dll!git_diff__oid_for_entry(git_oid * out, git_diff * d, const git_index_entry * src, unsigned short mode, const git_oid * update_match) Line 634	C
 	git2.dll!maybe_modified(git_diff_generated * diff, diff_in_progress * info) Line 847	C
 	git2.dll!handle_matched_item(git_diff_generated * diff, diff_in_progress * info) Line 1172	C
 	git2.dll!git_diff__from_iterators(git_diff * * out, git_repository * repo, git_iterator * old_iter, git_iterator * new_iter, const git_diff_options * opts) Line 1245	C
 	git2.dll!git_diff_index_to_workdir(git_diff * * out, git_repository * repo, git_index * index, const git_diff_options * opts) Line 1381	C
 	git2.dll!git_status_list_new(git_status_list * * out, git_repository * repo, const git_status_options * opts) Line 350	C
 	status.exe!main(int argc, char * * argv) Line 113	C
```

</p>
</details>

<details><summary>Callstack 5 (`path = "path/to/some/.gitattributes"`)</summary>
<p>

```>	git2.dll!lstat_w(wchar_t * path, p_stat * buf, unsigned char posix_enotdir) Line 303	C
 	git2.dll!p_stat(const char * path, p_stat * buf) Line 686	C
 	git2.dll!git_attr_file__load(git_attr_file * * out, git_repository * repo, git_attr_session * attr_session, git_attr_file_entry * entry, git_attr_file_source source, int(*)(git_repository *, git_attr_file *, const char *) parser) Line 140	C
 	git2.dll!git_attr_cache__get(git_attr_file * * out, git_repository * repo, git_attr_session * attr_session, git_attr_file_source source, const char * base, const char * filename, int(*)(git_repository *, git_attr_file *, const char *) parser) Line 231	C
 	git2.dll!push_attr_file(git_repository * repo, git_attr_session * attr_session, git_vector * list, git_attr_file_source source, const char * base, const char * filename) Line 448	C
 	git2.dll!push_one_attr(void * ref, const char * path) Line 472	C
 	git2.dll!git_path_walk_up(git_buf * path, const char * ceiling, int(*)(void *, const char *) cb, void * data) Line 514	C
 	git2.dll!collect_attr_files(git_repository * repo, git_attr_session * attr_session, unsigned int flags, const char * path, git_vector * files) Line 541	C
 	git2.dll!git_attr_get_many_with_session(const char * * values, git_repository * repo, git_attr_session * attr_session, unsigned int flags, const char * pathname, unsigned __int64 num_attr, const char * * names) Line 134	C
 	git2.dll!filter_list_check_attributes(const char * * * out, git_repository * repo, git_attr_session * attr_session, git_filter_def * fdef, const git_filter_source * src) Line 436	C
 	git2.dll!git_filter_list__load_ext(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, git_filter_options * filter_opts) Line 522	C
 	git2.dll!git_filter_list_load(git_filter_list * * filters, git_repository * repo, git_blob * blob, const char * path, git_filter_mode_t mode, unsigned int flags) Line 588	C
 	git2.dll!git_diff__oid_for_entry(git_oid * out, git_diff * d, const git_index_entry * src, unsigned short mode, const git_oid * update_match) Line 634	C
 	git2.dll!maybe_modified(git_diff_generated * diff, diff_in_progress * info) Line 847	C
 	git2.dll!handle_matched_item(git_diff_generated * diff, diff_in_progress * info) Line 1172	C
 	git2.dll!git_diff__from_iterators(git_diff * * out, git_repository * repo, git_iterator * old_iter, git_iterator * new_iter, const git_diff_options * opts) Line 1245	C
 	git2.dll!git_diff_index_to_workdir(git_diff * * out, git_repository * repo, git_index * index, const git_diff_options * opts) Line 1381	C
 	git2.dll!git_status_list_new(git_status_list * * out, git_repository * repo, const git_status_options * opts) Line 350	C
 	status.exe!main(int argc, char * * argv) Line 113	C
```

</p>
</details>

Along with many more very similar, e.g. walking up the path under `git_path_walk_up`. It leads to maybe ~10 `GetFileAttributesEx` calls per file, plus more depending on how deep the directory tree is.

It seems like maybe the `git_filter_*` functions are a common denominator, if this codepath to the `GetFileAttributesEx` call isn't typically expected to be hit.

**Comment by arguiot** [comment]
I saw this issue using [SwiftGit2](https://github.com/SwiftGit2/SwiftGit2/) and I posted this issue: https://github.com/SwiftGit2/SwiftGit2/issues/142 .

I profiled everything in Instruments, and the issue might come from `git_diff_index_to_workdir` at line `349` of `status.c`. I let the file here because you guys are more experienced than me using C and libgit2:

<img width="1396" alt="capture d ecran 2019-01-08 a 19 53 05" src="https://user-images.githubusercontent.com/18022260/50829309-052b2a00-137f-11e9-97d0-4dec6d5952b0.png">

[LibGit2 / SwiftGit2 status issue.trace.zip](https://github.com/libgit2/libgit2/files/2736589/LibGit2.SwiftGit2.status.issue.trace.zip)

**Comment by ethomson** [comment]
Hi @arguiot -

Neither git (the command line application) nor SourceTree for Mac actually use libgit2. 

```
Sourcetree:
	@rpath/Sparkle.framework/Versions/A/Sparkle (compatibility version 1.6.0, current version 1.20.0)
	/usr/lib/libicucore.A.dylib (compatibility version 1.0.0, current version 59.1.0)
	/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration (compatibility version 1.0.0, current version 963.50.8)
	@executable_path/../Frameworks/Growl.framework/Versions/A/Growl (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/CoreData.framework/Versions/A/CoreData (compatibility version 1.0.0, current version 851.0.0)
	/System/Library/Frameworks/Collaboration.framework/Versions/A/Collaboration (compatibility version 1.0.0, current version 80.0.0)
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 158.0.0)
	/System/Library/Frameworks/ServiceManagement.framework/Versions/A/ServiceManagement (compatibility version 1.0.0, current version 1205.50.76)
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 22.0.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 822.31.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 58286.51.6)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1452.23.0)
	/System/Library/Frameworks/QuickLook.framework/Versions/A/QuickLook (compatibility version 1.0.0, current version 0.0.0)
	@rpath/CocoaLumberjack.framework/Versions/A/CocoaLumberjack (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/Quartz.framework/Versions/A/Quartz (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/WebKit.framework/Versions/A/WebKit (compatibility version 1.0.0, current version 605.1.33)
	/System/Library/Frameworks/ScriptingBridge.framework/Versions/A/ScriptingBridge (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1452.23.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 1561.40.112)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1161.10.0)
	/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore (compatibility version 1.2.0, current version 1.11.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	@rpath/libswiftAppKit.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftCore.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftCoreData.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftCoreFoundation.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftCoreGraphics.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftCoreImage.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftDarwin.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftDispatch.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftFoundation.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftIOKit.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftMetal.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftObjectiveC.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftQuartzCore.dylib (compatibility version 1.0.0, current version 902.0.54)
	@rpath/libswiftXPC.dylib (compatibility version 1.0.0, current version 902.0.54)
```

I'd encourage you to open an issue on [Stack Overflow](https://stackoverflow.com/) with reproduction steps and a dump of the index file.

**Comment by arguiot** [comment]
@ethomson I maybe wasn’t clear enough: I’m not using a standard (built-in) Swift library but SwiftGit2 on iOS which uses libgit2 (and you can actually see it on the screenshot).
I already opened an issue there, but I strongly believe the issue is caused by libgit2 and not SwiftGit2.

**Comment by ethomson** [comment]
Sorry @arguiot, you were indeed clear, I meant this comment to be on #4679 instead.

**Comment by pks-t** [comment]
We had these problems in the past when it comes to diff generation, which I've fixed in d8896bda5 (diff_generate: avoid excessive stats of .gitattribute files, 2018-01-03). I suspect that we have the same problem in our status API, where we do not properly use the `git_attr_session`, and failure to do so will repeatedly invalidate the attributes cache.

**Comment by pks-t** [comment]
Interestingly enough, I can't really reproduce the problem with regards to the gitattributes file in a fresh clone of the Gentoo ports tree. My code:

```
#include "common.h"

int main(int argc, const char *argv[])
{
	git_status_options opts = GIT_STATUS_OPTIONS_INIT;
	git_status_list *status;
	git_repository *repo;
	size_t i;

	if (argc != 2)
		return -1;

	git_libgit2_init();

	opts.show = GIT_STATUS_SHOW_INDEX_AND_WORKDIR;
	opts.flags = GIT_STATUS_OPT_UPDATE_INDEX | 
		GIT_STATUS_OPT_INCLUDE_UNTRACKED |
		GIT_STATUS_OPT_RENAMES_HEAD_TO_INDEX |
		GIT_STATUS_OPT_EXCLUDE_SUBMODULES;

	check_lg2(git_repository_open_ext(&repo, argv[1], 0, NULL),
			"Could not open repository %s", argv[1]);
	check_lg2(git_status_list_new(&status, repo, &opts),
			"Could not get status", NULL);

	for (i = 0; i < git_status_list_entrycount(status); i++) {
		const git_status_entry *entry = git_status_byindex(status, i);
		(void)(entry);
	}

	git_libgit2_shutdown();

	return 0;
}
```

So this is essentially the same as the Git2Go code, but neither on current master nor on df4dfaa does it repeatedly check the ".gitattributes" files.

**Comment by pks-t** [comment]
What is interesting though is that we waste more than 30% of the time looking up whether an item is ignored or not. That's quite a lot of time.

**Comment by romkatv** [comment]
> I was looking at git2go/libgit2 to implement git status at shell $PS1 prompt...

(If you just want fast git status for your shell prompt, take a look at https://github.com/romkatv/powerlevel10k if you are using zsh; or at https://github.com/romkatv/gitstatus if using some other shell and willing to get your hands dirty.)

I had the same use-case in mind and I also was frustrated by how much slower libgit2 is compared to `git status` at detecting unstaged and untracked files.

After hours of profiling and experimenting I managed to speed up libgit2 by about 5x, making it faster than `git status` on every repo I've tried. I've made several optimizations.

1. Some of the hand-rolled string comparison functions in `util.c` are really hot. Replacing them with standard alternatives makes things faster. For example, I replaced `git__strcmp` with `strcmp`. There must be a reason why these functions were hand-rolled but I don't know what it is. I also streamlined `git__prefixcmp`, which had suboptimal codegen while being super hot.

2. When looking only for unstaged files, there is no need to check files in the workdir against ignore rules, but libgit2 does. These checks are expensive. After I eliminated them I saw a nice performance boost.

3. The biggest performance improvement came from using multiple threads to parallelize diffing (`git_diff_index_to_workdir` and `git_diff_tree_to_index`). To achieve this, I added `range_start` and `range_end` data members to `git_diff_options`. All diffing functions already delegate to diffing over iterators, and iterators have `start` and `end` to restrict their ranges. My `range_start` and `range_end` simply map to the iterators' `start` and `end`. This extension allowed me to shard a large repo into non-overlapping ranges and to diff ranges in parallel. The biggest problem is that `start` and `end` in iterators are somewhat broken. For example, if you run `git_diff_index_to_workdir` on the libgit2 repo itself and set `start` of both iterators to "tests/resources/crlf_data/windows_to_odb/autocrlf_true,safecrlf_true,text,eol_lf/" (keep `end` unset to scan to the end), it'll tell you that "tests/resources/crlf_data/windows_to_odb/autocrlf_true,safecrlf_true,text/all-crlf" is deleted. This string is indeed lexicographically greater than `start` we've specified, but workdir traversal doesn't happen in the lexicographical order of paths ("foo/bar" is visited before "foo,bar" by `filesystem_iterator` even though it's lexicographically greater), so it never sees this entry. I spent an hour trying to fix this but couldn't find a way to do this without performance cost, so I ended up with a workaround in my code. When sharding the index, I split it at points that act barriers for both types of iterators (index and workdir), meaning that no entry can appear before them in one traversal order and after them in another.

I should clarify that I used chromium repo as my main benchmark, and what I cared about is how fast I can detect whether my repo has staged, unstaged and untracked files. In chromium libgit2 spends a lot of time on checking files against ignore rules. In repos with fewer ignore rules optimizations (1) and (2) may be less important. (3) delivers reliable everywhere I've tried.

My patched libgit2 is here: https://github.com/romkatv/libgit2. Multi-threaded diffing code is here: https://github.com/romkatv/gitstatus/blob/master/src/git.cc.

**Comment by ethomson** [comment]
@romkatv that sounds very cool.  🤩💥✨  Are you planning to open a pull request?

**Comment by romkatv** [comment]
@ethomson I'm kinda trying to gauge whether I should. Is it something libgit2 folks care about? Is my code totally crazy? If I send a PR without first creating an issue, will it look an affront? I'm new here, cannot do like romans do when first time in rome.

**Comment by ethomson** [comment]
To answer a few of your questions:

> Some of the hand-rolled string comparison functions in util.c are really hot. Replacing them with standard alternatives makes things faster. For example, I replaced git__strcmp with strcmp. There must be a reason why these functions were hand-rolled but I don't know what it is. I also streamlined git__prefixcmp, which had suboptimal codegen while being super hot.

Because `strcmp` on some systems may decide to do a locale-sensitive search.  And this means taking a mutex on the locale to do so.  Moving to a native implementation is a significant performance win on Win32 with Visual Studio 2015 or better.  This seems like something we should both reevaluate (on win32) and avoid entirely (elsewhere).

> When looking only for unstaged files, there is no need to check files in the workdir against ignore rules, but libgit2 does. These checks are expensive. After I eliminated them I saw a nice performance boost.

Yes, this seems accurate.

**Comment by ethomson** [comment]
> @ethomson I'm kinda trying to gauge whether I should.

I think so!  We've got a lot of irons in the fire right now so we may not be able to get it merged immediately 😢 - especially if it's doing some crazy things that require deep thought or refactoring - but this would be interesting to me, indeed.

> Is it something libgit2 folks care about?

Yes.

> Is my code totally crazy?

Maybe.  :trollface: 

> If I send a PR without first creating an issue, will it look an affront? I'm new here, cannot do like romans do when first time in rome.

Nah, a PR is fine; no need to open an issue first.

My gut feeling is that there's a mix of easy wins (like the `strcmp` that you mentioned) and complex stuff (like threads), and we should carve off the easy wins immediately while we give more thought to the harder / more complex stuff.

**Comment by ethomson** [comment]
> Moving to a native implementation

Sorry, I meant to say _naive_.  😓

**Comment by romkatv** [comment]
@ethomson Got it. I'll send a PR after fixing whitespaces that got messed up (tabs vs spaces, sigh).

**Comment by ethomson** [comment]
> Got it. I'll send a PR after fixing whitespaces that got messed up (tabs vs spaces, sigh).

♥️♥️♥️

**Comment by romkatv** [comment]
@ethomson I sent you https://github.com/libgit2/libgit2/pull/5018 -- the smallest improvement both in terms of code size and impact. If this goes well, I'll send the other two. I'll already know the process by then, so things should go smoother than if I would send all three PRs right away (or even worse -- one PR with 3 independent improvements in it).

**Comment by romkatv** [comment]
I've implemented several more optimizations and now my version of `git status` is about 10 times faster than `lg2 status` and 2.5 times faster than the regular `git status`. This is in chromium git repo where I've done most benchmarking.

The biggest thing I did since my last comment is to extend gitignore optimization I've done previously. Now, when I run my `git status` on a clean repo that has just been cloned, not only there is no checking against gitignore rules but gitignore files don't even get parsed. When a file/directory is encountered that exists in the workdir but doesn't exist in the index, only then relevant gitignore files are parsed and the file/directory is checked against the rules. More generally, the minimum required set of gitignore files is parsed and the minimum required number of files are checked against the rules in order to generate correct diff. This gives gives about 65% speedup.

I've optimized small things here and there. For example, a small optimization in `parse_mode` gave 6% performance boost.

I've also done some shady things that make sense in my environment but not generally. For example, I added an option to disable index filepath validation when reading an index. Since my code is just doing `git status` and is never writing anything to the workdirectory, it doesn't need this validation.

Since I use libgit2 from multiple threads in a way in which it wasn't supposed to be used (I do index-to-workdir scan in parallel), I bumped into a few synchronization bugs (or not bugs, depending on your point of view) that I've fixed. Some of them were clearly in the code that is supposed to work in concurrent setting (e.g., attr cache), while some other was probably not thread safe by design.

Since my first PR isn't moving very fast, if anyone is interested in 2-3 times faster `git status`, you can find my patched libgit2 in https://github.com/romkatv/libgit2 and the code that uses it in https://github.com/romkatv/gitstatus/. This is all part of [Pure Power](https://www.reddit.com/r/zsh/comments/b45w6v/pure_power_a_beautiful_theme_with_a_powerful_punch/).

P.S.

Kudos to the developers of libgit2. The code is a pleasure to read and modify. Elegant, clean and simple.

**Comment by tiennou** [comment]
Hi @romkatv! Sorry for the long delay in reviewing, I had been planning to get a close look at your changes, but am short on time. The fact that there's no actual PR is also somewhat limiting 😉. Hence, after having taking a look at your branch, I've pushed a bunch of trivial rewrites/splits to have more separated concerns. I went through the tabs->space issues, and tweaked a few things for self-review reasons*. Now, I have more in-depth comments about those, and I'd prefer to keep the authorship around, so if you filed those as PR here, that's allow us to keep tabs on them.

So there's :
- https://github.com/libgit2/libgit2/compare/master...tiennou:romkatv/iterator-skip-ignores, which has the fix for skipping ignores when not diffing workdirs. Note that I've reversed the semantics of the flags because it made more sense from a backward-compatiibility point of view, and also limits the amount of churn in the testsuite.
- https://github.com/libgit2/libgit2/compare/master...tiennou:romkatv/diff-sharding, which has everything sharding-related.
- https://github.com/libgit2/libgit2/compare/master...tiennou:romkatv/perf, which is what's left, so mostly performance improvements.

Many thanks for taking the time to prod at the status code!

* we have a convention of prefixing the subsystem in commit messages. I've added it to some of the earlier commits but stopped after doing the splits.

**Comment by romkatv** [comment]
@tiennou Wow, thanks! Going through my messy commits must have been quite a bit of work.

Can I propose that your file the PRs and list me as a reviewer? I can annotate the changes through code review comments to make it clear why each one was made.

**iterator-skip-ignores** looks great. The reversed semantics of flags is definitely better than what I had.

**diff-sharding** pulls in some changes unrelated to sharding. For example, `GIT_OPT_DISABLE_INDEX_CHECKSUM_VERIFICATION` and `GIT_OPT_DISABLE_INDEX_FILEPATH_VALIDATION`. I'm not sure if this is intended.

Some commits in **diff-sharding** and **perf** will need changes to make them portable. For example, I used gcc builtins, which likely won't work with all compilers you support. There are a few commits where I took shortcuts. For example, I renamed `git_index_iterator_new` to `git_index_iterator_new_testonly` because I didn't want to spend time fixing it; since it's only used in tests it was sufficient to rename it and be certain I didn't break code in libgit2 on which gitstatusd relies.

It'll definitely be easier if I could communicate this sort of information in a code review.

**Comment by danzimm** [comment]
Hey, just wanted to bump this thread as I've ran into the same issues described above. Could I help in anyway to get the PRs mentioned above pushed forward?

Cheers

**Comment by ethomson** [comment]
> Hey, just wanted to bump this thread as I've ran into the same issues described above. Could I help in anyway to get the PRs mentioned above pushed forward?

Sure!  @tiennou was kind enough to create those PRs and @romkatv gave some really helpful feedback on them that needs to be incorporated.

First step would be rebasing those PRs as they exist on master and make sure that they build.  After that, addressing @romkatv's feedback would be :tada:.

It might be worth breaking them up into smaller-still chunks, that certainly helps with reviewability (and mergeability).  So instead of taking the whole PR as it exists, you might want to cleave off a functional area like was done in #5018.  For example, reducing the calls to error setting in non error cases would likely be a big win.

**Comment by twa16** [comment]
@danzimm Did you end up digging into this? This is going to be an issue for me in a few weeks so I should be able to dig into this then if you are not able to.

**Comment by pnordahl** [comment]
I can confirm that this is an issue for me as well, using https://github.com/rust-lang/git2-rs. Extremely slow performance of `statuses` more or less forcing me to shell out to `git status` instead; it's roughly a 20x slowdown.

It would be great if this could be improved to performance parity with `git`!

**Comment by personalizedrefrigerator** [comment]
I would like to link this issue to #5808!

**Comment by voliva** [comment]
Also effected by this on large repositories

Not sure if this will help or distract, but I tried circumventing `git_status_list` by using `git_diff_index_to_workdir` and `git_diff_tree_to_index`, but it's equally slow. `git_diff_tree_to_index` is fast, but `git_diff_index_to_workdir` takes the same amount of time as `git_status_list` approximately.

**Comment by SuperBo** [comment]
Any progress on this issue?
